library/x509.c

```diff
@@ -840,32 +853,95 @@ int mbedtls_x509_dn_gets(char *buf, size_t size, const mbedtls_x509_name *dn)
+                          (name->val.tag != MBEDTLS_ASN1_PRINTABLE_STRING) &&
+                          (name->val.tag != MBEDTLS_ASN1_IA5_STRING);
 
-        if (ret == 0) {
+        if ((ret = mbedtls_oid_get_attr_short_name(&name->oid, &short_name)) == 0) {
             ret = mbedtls_snprintf(p, n, "%s=", short_name);
         } else {
-            ret = mbedtls_snprintf(p, n, "\?\?=");
+            if ((ret = mbedtls_oid_get_numeric_string(p, n, &name->oid)) > 0) {
+                n -= ret;
+                p += ret;
+                ret = mbedtls_snprintf(p, n, "=");
+                print_hexstring = 1;
+            } else if (ret == MBEDTLS_ERR_OID_BUF_TOO_SMALL) {
+                return MBEDTLS_ERR_X509_BUFFER_TOO_SMALL;
+            } else {
+                ret = mbedtls_snprintf(p, n, "\?\?=");
+            }
         }
         MBEDTLS_X509_SAFE_SNPRINTF;
 
-        for (i = 0, j = 0; i < name->val.len; i++, j++) {
-            if (j >= sizeof(s) - 1) {
-                return MBEDTLS_ERR_X509_BUFFER_TOO_SMALL;
-            }
+        if (print_hexstring) {
+#if defined(MBEDTLS_ASN1_WRITE_C)
+            s[0] = '#';
 
-            c = name->val.p[i];
-            // Special characters requiring escaping, RFC 1779
-            if (c && strchr(",=+<>#;\"\\", c)) {
+            asn1_len_p = asn1_tag_len_buf + 10;
+            if ((ret = mbedtls_asn1_write_len(&asn1_len_p, asn1_tag_len_buf, name->val.len)) < 0) {
+                return MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+            }
+            asn1_len_size = ret;
+            if ((ret = mbedtls_asn1_write_tag(&asn1_len_p, asn1_tag_len_buf, name->val.tag)) < 0) {
+                return MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+            }
+            asn1_tag_size = ret;
+            asn1_tag_len_buf_start = 10 - asn1_len_size - asn1_tag_size;
+            for (i = 0, j = 1; i < asn1_len_size + asn1_tag_size; i++) {
                 if (j + 1 >= sizeof(s) - 1) {
                     return MBEDTLS_ERR_X509_BUFFER_TOO_SMALL;
                 }
-                s[j++] = '\\';
+                c = asn1_tag_len_buf[asn1_tag_len_buf_start+i];
+                lowbits = (c & 0x0F);
+                highbits = c >> 4;
+                s[j++] = nibble_to_hex_digit(highbits);
+                s[j++] = nibble_to_hex_digit(lowbits);
             }
-            if (c < 32 || c >= 127) {
-                s[j] = '?';
-            } else {
-                s[j] = c;
+            for (i = 0; i < name->val.len; i++) {
+                if (j + 1 >= sizeof(s) - 1) {
+                    return MBEDTLS_ERR_X509_BUFFER_TOO_SMALL;
+                }
+                c = name->val.p[i];
+                lowbits = (c & 0x0F);
+                highbits = c >> 4;
+                s[j++] = nibble_to_hex_digit(highbits);
+                s[j++] = nibble_to_hex_digit(lowbits);
+            }
+#else
+            return MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE;
```

Comment:
Would you think it would be better to change this line to `MBEDTLS_ERR_X509_INVALID_NAME` or to just make the whole function not work at all if ASN1_WRITE_C is disabled?

Reply:
Hmm, good question. What would happen with such an input previously? And I note that this is a "get" function - it seems weird that its behaviour depends upon a "write" configuration parameter. Maybe the correct fix here is to allow the needed functions even if we don't have ASN1_WRITE_C?

@davidhorstmann-arm you've spent more time in the ASN.1 modules than I have - do you have an opinion here?

Reply:
> allow the needed functions even if we don't have ASN1_WRITE_C?

I don't understand what this code is doing, but yes, this seems like the right thing.

Why is parsing code calling write functions in the first place? This code needs comments explaining what's going on.

Reply:
> do you have an opinion here?

I think we should leave it as-is. Let me give a quick background of what is going on here:

In cases where a Relative Distinguished Name is not a string, RFC 4514 allows us to output the RDN as a string in the form:
```
<OID number>=#<hex encoding of raw der>
```
For example:
```
1.3.6.1.4.1.1466.0=#04024869
```
We need `MBEDTLS_ASN1_WRITE_C` to encode the DER data that gets put after the `#`, because we don't store it in its raw form but unpacked in a struct.

Most RDNs are strings and don't need this alternate form, so ideally we should allow users to use this function without needing `MBEDTLS_ASN1_WRITE_C`. When we encounter a name that is not a string, we return `MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE`, as in "the feature of printing non-string RDNs is unavailable, because you didn't enable `MBEDTLS_ASN1_WRITE_C`".

However, I'm not sure whether this aligns with the way we use `FEATURE_UNAVAILABLE` normally, as this is not a clearly delineated "X509 Feature". Do you think that `INVALID_NAME` would be a better choice?

In any case, it looks like we at least need some code comments here, explaining what is going on and why we need `MBEDTLS_ASN1_WRITE_C`. Additionally, we could do with a note in the doxygen comment in the header, something along the lines of:
```doxygen
\note           Converting names with arbitrary non-string RDNs
                requires MBEDTLS_ASN1_WRITE_C to be enabled.
```

Reply:
Thanks, that matches with what my understanding was.

However, if we're going to enhance the parsing of ASN.1 data, having it depend on a "WRITE" configuration value just seems wrong.

> Most RDNs are strings and don't need this alternate form, so ideally we should allow users to use this function without needing `MBEDTLS_ASN1_WRITE_C`.

The issue is that the type of RDN that the user is going to pass to this function isn't necessarily known at compile time, so having it depend on a compile-time value is sub-optimal.

> we don't store it in its raw form but unpacked in a struct.

That's our choice/problem, so shouldn't make the user's life harder.

> we could do with a note in the doxygen comment in the header, something along the lines of ...

That would have made review easier :) but I don't think it's the right approach here.

Reply:
So I think those functions needed by this function need to be enabled by `MBEDTLS_X509_USE_C`as well as `MBEDTLS_ASN1_WRITE_C`

Reply:
> The issue is that the type of RDN that the user is going to pass to this function isn't necessarily known at compile time, so having it depend on a compile-time value is sub-optimal.

I think the easiest way to do this would be to duplicate the functionality of the `asn1_write` functions locally in this file. They are quite simple, so this is definitely doable without too much risk. Would this be a workable solution?

Reply:
Sorry, I disagree: `mbedtls_asn1_write_len()` alone is 63 lines of C, and 186 bytes of code in Thumb2 - that's not something we should be duplicating at all

Reply:
Okay that's fair enough, I suppose they are larger than I remembered.

Ah I see now what you mean - apologies, I completely misunderstood. You propose that we should move the definitions of `mbedtls_asn1_write_tag()` and `mbedtls_asn1_write_len()` so that they are included even without `MBEDTLS_ASN1_WRITE_C`?

This solution seems fine to me, and I agree that it's the best thing we can do in the circumstances.

Reply:
> You propose that we should move the definitions [...] so that they are included even without MBEDTLS_ASN1_WRITE_C (but with MBEDTLS_X509_USE_C).

Yes - they now become part of the "core" X.509 code, rather than functions only needed for WRITE.

Some careful testing will be necessary to ensure that the library still builds under various combination of config options